### PR TITLE
Reconciler operations for BPF maps

### PR DESCRIPTION
**Pull request by joamaki**

### CODE REVIEW

File: pkg/bpf/mapreconciler/example/map.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Enabled implements mapreconciler.Map.
+func (*exampleMap) Enabled() bool {
+	return true
+}
+
+// Iterate implements mapreconciler.Map.
+func (*exampleMap) Iterate() mapreconciler.Iterator[ExampleKey, ExampleValue] {
+	panic("unimplemented")
```
Comment (joamaki) [inline pkg/bpf/mapreconciler/example/map.go]: TODO: Implement this

File: pkg/bpf/mapreconciler/ops.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			}
+		}
+	}
+	errs = append(errs, mapIter.Err())
+	return errors.Join(errs...)
+}
+
+// Update implements reconciler.Operations.
+func (ops *mapOps[KV, K, V]) Update(ctx context.Context, txn statedb.ReadTxn, entry KV) (bool, error) {
+	// TODO: Not cheap to check if this changed the value. Returning true will mess up the
```
Comment (joamaki) [inline pkg/bpf/mapreconciler/ops.go]: Figure out a solution to this.

Comment (joamaki) [reply pkg/bpf/mapreconciler/ops.go]: could do:
```
Update(ctx context.Context, txn statedb.ReadTxn, obj Obj, changed *bool) error
```
If `changed` is `nil` then `Update` doesn't need to check if it's changed.
For BPF map reconciler we'd have `Update` do a `Lookup` and compare if `changed != nil`. Full reconciler would then pass this in.
It's little bit ugly though.

File: pkg/statedb/reconciler/example/example
Comment (pippolo84) [inline pkg/statedb/reconciler/example/example:1]: This is probably a leftover. Maybe we should add a .gitignore here and in pkg/hive/example.

File: pkg/statedb/reconciler/incremental.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+	if len(round.errs) > 0 {
+		return newRevision, watch, fmt.Errorf("incremental: %w", joinErrors(round.errs))
+	}
+	return newRevision, watch, nil
+}
+
+func (round *incrementalRound[Obj]) single(ops Operations[Obj], labels prometheus.Labels) statedb.Revision {
+	// Function to process the new&changed objects
+	process := func(obj Obj, rev statedb.Revision, status Status) error {
```
Comment (pippolo84) [inline pkg/statedb/reconciler/incremental.go]: This seems equal to the `process` function at line 253, maybe better to make it a package-level function

File: pkg/bpf/ops/ops.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		} else {
+			*changed = !ops.equalValue(value, entry)
+		}
+		if *changed {
+			return ops.m.Put(entry.Key(), entry.Value())
+		}
+		return nil
+	} else {
+		return ops.m.Put(entry.Key(), entry.Value())
+	}
```
Comment (pippolo84) [inline pkg/bpf/ops/ops.go]: Nit: this feels a bit more readable to me if we use early-return:
```
	if changed == nil {
		return ops.m.Put(entry.Key(), entry.Value())
	}

	var value []byte
	err := ops.m.Lookup(entry.Key(), &value)
	if err != nil {
		if !errors.Is(err, cilium_ebpf.ErrKeyNotExist) {
			return err
		}
		*changed = true
	} else {
		*changed = !ops.equalValue(value, entry)
	}
	if *changed {
		return ops.m.Put(entry.Key(), entry.Value())
	}
	return nil
```

**Review by pippolo84 — APPROVED** [review]
Great! :100: 
Just some few (minor) comments here and there, it's ok to merge after addressing them.

File: pkg/statedb/reconciler/example/example
Comment (joamaki) [reply pkg/statedb/reconciler/example/example:1]: Fixed.

File: pkg/statedb/reconciler/incremental.go (hunk 3 above)
Comment (joamaki) [reply pkg/statedb/reconciler/incremental.go]: Dedup'd.

File: pkg/bpf/ops/ops.go (hunk 4 above)
Comment (joamaki) [reply pkg/bpf/ops/ops.go]: Yep reads better, fixed.

**Comment by joamaki** [comment]
Merged into feature branch.

**Pull request closed without merging**